gt/communication/ghex_comm_mt.hpp

```diff
@@ -0,0 +1,179 @@
+        }()}
+    , m_first{(int)global_resolution.x * m_coords[0] / m_dims[0],
+              (int)global_resolution.y * m_coords[1] / m_dims[1], 
+              0}
+    , m_last{ (int)global_resolution.x * (m_coords[0] + 1) / m_dims[0] - 1,
+              (int)global_resolution.y * (m_coords[1] + 1) / m_dims[1] - 1,
+              (int)global_resolution.z - 1}
+    , m_hg{std::array<int, 3>{0, 0, 0},
+           std::array<int, 3>{(int)global_resolution.x - 1, (int)global_resolution.y - 1, (int)global_resolution.z - 1}, 
+           halo}
+    , m_global_resolution{global_resolution.x, global_resolution.y}
+    {
+        m_domains.reserve(num_sub_domains);
+
+        auto div = divide_domain(num_sub_domains, std::array<int,2>{m_last[0]-m_first[0]+1,m_last[1]-m_first[1]+1});
+
+        int i = 0;
+        int y = m_first[1];
+        for (const auto& dy : div[1])
+        {
+            int x = m_first[0];
+            for (const auto& dx : div[0])
+            {
+                //std::cout << "[(" << x << ", " << y << "), (" << x+dx-1 << ", " << y+dy-1 << ")]" << std::endl;
+                m_domains.push_back( local_domain{ 
+                    m_rank*num_sub_domains+i,
+                    coordinate_type{x,y,m_first[2]},
+                    coordinate_type{x+dx-1,y+dy-1,m_last[2]} });
+                x+=dx;
+                ++i;
+            }
+            y+=dy; 
+        }
+        
+        m_patterns.reset( new patterns_type{
+            ::gridtools::ghex::make_pattern<::gridtools::ghex::structured::grid>(m_comm_cart, m_hg, m_domains)});
+    }
+
+    grid_mt(const grid_mt&) = delete;
+    grid_mt(grid_mt&&) = default;
+    ~grid_mt() { if (!m_moved) MPI_Comm_free(&m_comm_cart); }
+
+public: // member functions
+
+    sub_grid operator[](unsigned int i) const
+    {
+        const auto& dom = m_domains[i];
+
+        return { dom.domain_id(),
+                 m_comm_cart,
+                 comm_obj_ptr_t{new comm_obj_type{::gridtools::ghex::make_communication_object<patterns_type>()}},
+                 m_patterns.get(),
+                 m_global_resolution, 
+                 {(std::size_t)dom.first()[0],
+                  (std::size_t)dom.first()[1]},
+                 {(std::size_t)(dom.last()[0]-dom.first()[0]+1),
+                  (std::size_t)(dom.last()[1]-dom.first()[1]+1),
+                  (std::size_t)(dom.last()[2]-dom.first()[2]+1)}};
+
+    }
+
+protected: // implementation
+};
+
+std::function<void(storage_t &)>
+comm_halo_exchanger(grid_mt::sub_grid &g, storage_t::storage_info_t const &sinfo)
+{
+    auto co_ptr          = g.co.get();
+    auto patterns_ptr    = g.patterns;
+    const auto domain_id = g.domain_id;
+    
+    return [co_ptr, patterns_ptr, domain_id](const storage_t &storage) mutable
+    {
+        auto &co = *co_ptr;
+        auto &patterns = *patterns_ptr;
+        auto field = ::gridtools::ghex::wrap_gt_field(domain_id, storage);
+
+        #ifdef __CUDACC__
+        cudaDeviceSynchronize();
```

Comment:
general question: should we put the `cudaDeviceSynchronize` here or are we then calling this function too often?

Reply:
we absolutely need it here! ghex does not use stream 0 so stuff will not be synchronized